Windows/Cygwin: hygienic check fails for some packages

It looks like an [old issue](https://github.com/ocaml/opam-repository/issues/6224) has popped up again in `ocamlbuild` `0.14.1`.
For some packages on Windows/Cygwin (Mingw toolchain), the hygienic check fails, e.g., for `mtime` `1.3.0`:
```
mkdir 'C:\Opt\OCaml64\home\kayuwe.kirstein\.opam\4.13.1+mingw64\.opam-switch\build\mtime.1.3.0\_build'
Das System kann den angegebenen Pfad nicht finden.
''ocamlfind ocamlopt unix.cmxa -I 'C:/Opt/OCaml64/home/kayuwe.kirstein/.opam/4.13.1+mingw64/lib\ocamlbuild' 'C:/Opt/OCaml64/home/kayuwe.kirstein/.opam/4.13.1+mingw64/lib\ocamlbuild/ocamlbuildlib.cmxa' -linkpkg myocamlbuild.ml 'C:/Opt/OCaml64/home/kayuwe.kirstein/.opam/4.13.1+mingw64/lib\ocamlbuild/ocamlbuild.cmx' -o myocamlbuild.exe
SANITIZE: a total of 3 files that should probably not be in your source tree
  has been found. A script shell file
  "C:\\Opt\\OCaml64\\home\\kayuwe.kirstein\\.opam\\4.13.1+mingw64\\.opam-switch\\build\\mtime.1.3.0\\_build/sanitize.sh"
  is being created. Check this script and run it to remove unwanted files or
  use other options (such as defining hygiene exceptions or using the
  -no-hygiene option).
IMPORTANT: I cannot work with leftover compiled files.
ERROR: Leftover object files:
  File myocamlbuild.o in _build has suffix .o
ERROR: Leftover OCaml compilation files:
  File myocamlbuild.cmi in _build has suffix .cmi
  File myocamlbuild.cmx in _build has suffix .cmx
Exiting due to hygiene violations.
pkg.ml: [ERROR] cmd ["ocamlbuild" "-use-ocamlfind" "-classic-display" "-j" "4" "-tag" "debug"
     "-build-dir" "_build" "opam" "pkg/META" "CHANGES.md" "LICENSE.md"
     "README.md" "src/mtime.a" "src/mtime.cmxs" "src/mtime.cmxa"
     "src/mtime.cma" "src/mtime.cmx" "src/mtime.cmi" "src/mtime.mli"
     "src/mtime_top.a" "src/mtime_top.cmxs" "src/mtime_top.cmxa"
     "src/mtime_top.cma" "src/mtime_top.cmx" "src/mtime_top_init.ml"
     "doc/index.mld" "src-os/mtime_clock.a" "src-os/mtime_clock.cmxs"
     "src-os/mtime_clock.cmxa" "src-os/mtime_clock.cma"
     "src-os/mtime_clock.cmx" "src-os/mtime_clock.cmi"
     "src-os/mtime_clock.mli" "src-os/dllmtime_clock_stubs.dll"
     "src-os/libmtime_clock_stubs.a" "test-os/min_os.ml"]: exited with 1
```
I am not sure whether the first error message on not finding the build folder is a serious failure.
Executing `sanitize.sh` does not help, I guess `opam install mtime` regenerates the object files every time.

Observed on Windows 10, Cygwin/Mingw (Ocaml for Windows) for OCaml `4.13.1+mingw64`.
My current workaround is downgrading `ocamlbuild` to `0.14.0`.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -16,6 +16,7 @@ jobs:
       matrix:
         os:
           - ubuntu-latest
+          - windows-latest
         ocaml-compiler:
           - "4.08"
           - "4.09"
@@ -66,7 +67,7 @@ jobs:
 
       - run: opam install . --with-test --deps-only
 
-      - run: VERBOSE=1 opam exec -- make test
+      - run: opam exec -- make test
 
       - run: OCAMLBUILD=$(pwd)/ocamlbuild.native opam exec -- make -C bootstrap
 

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -25,15 +25,14 @@ jobs:
           - "4.12"
           - "4.13"
           - "4.14"
-          - "5.0"
-          - "5.1"
-          - "5.2"
+          - "5.0.0"
+          - "5.1.1"
+          - "5.2.0"
 
     runs-on: ${{ matrix.os }}
 
     steps:
       - name: Set git to use LF
-        if: matrix.ocaml-compiler < 5.2
         run: |
           git config --global core.autocrlf false
           git config --global core.eol lf
@@ -43,10 +42,22 @@ jobs:
         uses: actions/checkout@v4
 
       - name: Set-up OCaml ${{ matrix.ocaml-compiler }}
-        if: runner.os == 'Windows'
+        if: runner.os == 'Windows' && ! startsWith(matrix.ocaml-compiler, '5')
         uses: ocaml/setup-ocaml@v2
         with:
           ocaml-compiler: ${{ matrix.ocaml-compiler }}
+          opam-repositories: |
+            default: https://github.com/ocaml-opam/opam-repository-mingw.git#sunset
+            opam: https://github.com/ocaml/opam-repository.git
+          dune-cache: true
+          opam-depext: ${{ !matrix.skip-test }}
+          opam-depext-flags: --with-test
+
+      - name: Set-up OCaml ${{ matrix.ocaml-compiler }}
+        if: runner.os == 'Windows' && startsWith(matrix.ocaml-compiler, '5')
+        uses: ocaml/setup-ocaml@v2
+        with:
+          ocaml-compiler: ocaml-variants.${{ matrix.ocaml-compiler }}+options,ocaml-option-mingw
           opam-repositories: |
             dra27: https://github.com/dra27/opam-repository.git#windows-5.0
             default: https://github.com/ocaml-opam/opam-repository-mingw.git#sunset
@@ -70,13 +81,14 @@ jobs:
       - run: opam exec -- make test
 
       - run: OCAMLBUILD=$(pwd)/ocamlbuild.native opam exec -- make -C bootstrap
+        if: runner.os != 'Windows'
 
       - run: opam exec -- make distclean
-      - run: opam exec -- opam pin add -n -k path ocamlbuild .
+      - run: opam exec -- opam pin add -n ocamlbuild .
       - run: opam exec -- opam install -v ocamlbuild
       - run: opam exec -- opam install -v mtime.1.0.0 # this tests topkg, with stub libraries
       - run: opam exec -- opam install -v inotify.2.3 # this tests oasis, with stub libraries
-        if: matrix.ocaml-compiler < 5
+        if: (! startsWith(matrix.ocaml-compiler, '5')) && runner.os != 'Windows'
       - run: opam exec -- opam install -v cpuid.0.1.1 # this tests the ocb-stubblr plugin
       - run: opam exec -- opam install -v shcaml.0.2.1 # this tests the cppo plugin
-        if: matrix.ocaml-compiler < 5
+        if: (! startsWith(matrix.ocaml-compiler, '5')) && runner.os != 'Windows'

--- a/Makefile
+++ b/Makefile
@@ -30,7 +30,7 @@ OCAMLLEX  ?= ocamllex
 endif
 
 CP        ?= cp
-COMPFLAGS ?= -w @14 -w +L -w +R -w +Z -I src -I plugin-lib -I bin -I +unix -safe-string -bin-annot -strict-sequence
+COMPFLAGS ?= -w @14@29 -w +L -w +R -w +Z -I src -I plugin-lib -I bin -I +unix -safe-string -bin-annot -strict-sequence
 LINKFLAGS ?= -I +unix -I src
 
 PACK_CMO= $(addprefix src/,\

--- a/Makefile
+++ b/Makefile
@@ -225,7 +225,7 @@ endif
 # Testing
 
 test-%: testsuite/%.ml all
-	cd testsuite && ocaml $(CURDIR)/$<
+	cd testsuite && ocaml "../$<"
 
 test: test-internal test-findlibonly test-external
 

--- a/Makefile
+++ b/Makefile
@@ -225,7 +225,7 @@ endif
 # Testing
 
 test-%: testsuite/%.ml all
-	cd testsuite && ocaml "../$<"
+	cd testsuite && ocaml -w @14@29 "../$<"
 
 test: test-internal test-findlibonly test-external
 

--- a/Makefile
+++ b/Makefile
@@ -30,7 +30,7 @@ OCAMLLEX  ?= ocamllex
 endif
 
 CP        ?= cp
-COMPFLAGS ?= -w +L -w +R -w +Z -I src -I plugin-lib -I bin -I +unix -safe-string -bin-annot -strict-sequence
+COMPFLAGS ?= -w @14 -w +L -w +R -w +Z -I src -I plugin-lib -I bin -I +unix -safe-string -bin-annot -strict-sequence
 LINKFLAGS ?= -I +unix -I src
 
 PACK_CMO= $(addprefix src/,\

--- a/configure.make
+++ b/configure.make
@@ -92,10 +92,10 @@ Makefile.config:
 src/ocamlbuild_config.ml:
 	(echo "(* This file was generated from ../configure.make *)"; \
 	echo ;\
-	echo 'let bindir = "$(OCAMLBUILD_BINDIR)"'; \
-	echo 'let libdir = "$(OCAMLBUILD_LIBDIR)"'; \
-	echo 'let ocaml_libdir = "$(abspath $(OCAML_LIBDIR))"'; \
-	echo 'let libdir_abs = "$(abspath $(OCAMLBUILD_LIBDIR))"'; \
+	echo 'let bindir = {|$(OCAMLBUILD_BINDIR)|}'; \
+	echo 'let libdir = {|$(OCAMLBUILD_LIBDIR)|}'; \
+	echo 'let ocaml_libdir = {|$(abspath $(OCAML_LIBDIR))|}'; \
+	echo 'let libdir_abs = {|$(abspath $(OCAMLBUILD_LIBDIR))|}'; \
 	echo 'let ocaml_native = $(OCAML_NATIVE)'; \
 	echo 'let ocaml_native_tools = $(OCAML_NATIVE_TOOLS)'; \
 	echo 'let supports_shared_libraries = $(SUPPORTS_SHARED_LIBRARIES)';\

--- a/plugin-lib/ocamlbuild_unix_plugin.ml
+++ b/plugin-lib/ocamlbuild_unix_plugin.ml
@@ -48,6 +48,14 @@ let at_exit_once callback =
   end
 
 let run_and_open s kont =
+  let s =
+    (* Be consistent! My_unix.run_and_open uses My_std.sys_command and
+       sys_command uses bash. *)
+    if Sys.win32 then
+      "bash --norc -c " ^ Filename.quote s
+    else
+      s
+  in
   let ic = Unix.open_process_in s in
   let close () =
     match Unix.close_process_in ic with

--- a/src/command.ml
+++ b/src/command.ml
@@ -294,8 +294,7 @@ let execute_many ?(quiet=false) ?(pretend=false) cmds =
                         let rc = sys_command (cmd ^ " 2>&1") in
                         if rc <> 0 then begin
                           if not quiet then
-                            eprintf "Exit code %d while executing this \
-                                    command:@\n%s" rc cmd;
+                            eprintf "Command exited with code %d." rc;
                           raise (Exit_with_code rc)
                         end
                       end cmds;

--- a/src/command.ml
+++ b/src/command.ml
@@ -289,7 +289,9 @@ let execute_many ?(quiet=false) ?(pretend=false) cmds =
                     begin try
                       List.iter begin fun action ->
                         let cmd = action () in
-                        let rc = sys_command cmd in
+                        (* Redirect stderr to stdout to match the
+                           behavior of My_unix.execute_many *)
+                        let rc = sys_command (cmd ^ " 2>&1") in
                         if rc <> 0 then begin
                           if not quiet then
                             eprintf "Exit code %d while executing this \

--- a/src/display.ml
+++ b/src/display.ml
@@ -370,7 +370,7 @@ let event di ?(pretend=false) command target tags =
           if di.di_log_level >= 2 then Format.fprintf di.di_formatter "[cache hit] %s\n%!" command
         end
       else
-        (if di.di_log_level >= 1 then Format.fprintf di.di_formatter "%s\n%!" command)
+        (if di.di_log_level >= 1 then Format.fprintf di.di_formatter "+ %s\n%!" command)
   | Sophisticated ds ->
       set_target_sophisticated ds target tags pretend;
       update_sophisticated ds

--- a/src/lexers.mll
+++ b/src/lexers.mll
@@ -122,7 +122,7 @@ and conf_lines dir source = parse
       {
         let bexpr =
           try Glob.parse ?dir k
-          with exn -> error source lexbuf "Invalid globbing pattern %S" k (Printexc.to_string exn)
+          with exn -> error source lexbuf "Invalid globbing pattern %S" k
         in
         sublex (count_lines lexbuf) s1; sublex (count_lines lexbuf) s2;
         let v1 = conf_value empty source lexbuf in

--- a/src/ocaml_compiler.ml
+++ b/src/ocaml_compiler.ml
@@ -125,9 +125,9 @@ let prepare_compile build ml =
           "Warning: Failed to build the module %s requested by ocamldep."
           name;
         if not (!Options.recursive || Options.ocamlbuild_project_heuristic ())
-        then Log.at_failure ~name:"a module failed to build,
-           while recursive traversal was disabled by fragile heuristic;
-           hint that having a _tags or myocamlbuild.ml would maybe solve
+        then Log.at_failure ~name:"a module failed to build, \
+           while recursive traversal was disabled by fragile heuristic;\n\
+           hint that having a _tags or myocamlbuild.ml would maybe solve \
            the build error"
           (fun `Error ->
             eprintf "Hint:@ Recursive@ traversal@ of@ subdirectories@ \

--- a/src/ocaml_specific.ml
+++ b/src/ocaml_specific.ml
@@ -109,14 +109,12 @@ rule "ocaml: mlpack & cmo* & cmi -> cmo"
   ~doc:"If foo.mlpack contains a list of capitalized module names, \
   the target foo.cmo will produce a packed module containing \
   those modules as submodules. You can also have a foo.mli file \
-  to restrict the interface of the resulting module.
-
-\
+  to restrict the interface of the resulting module.\n\
+  \n\
   Warning: to produce a native foo.cmx out of a foo.mlpack, you must \
   manually tag the included compilation units with for-pack(foo). \
-  See the documentation of the corresponding rules for more details.
-
-\
+  See the documentation of the corresponding rules for more details.\n\
+  \n\
   The modules named in the .mlpack \
   will be dynamic dependencies of the compilation action. \
   You cannot give the .mlpack the same name as one of the module \
@@ -132,10 +130,9 @@ rule "ocaml: ml & cmi -> d.cmo"
   ~prod:"%.d.cmo"
   ~deps:["%.mli"(* This one is inserted to force this rule to be skiped when
     a .ml is provided without a .mli *); "%.ml"; "%.ml.depends"; "%.cmi"]
-  ~doc:"The foo.d.cmo target compiles foo.ml with the 'debug' tag enabled (-g).\
-        See also foo.d.byte.
-
-\
+  ~doc:"The foo.d.cmo target compiles foo.ml with the 'debug' tag enabled (-g). \
+        See also foo.d.byte.\n\
+        \n\
         For technical reason, .d.cmx and .d.native are not yet supported, \
         so you should explicitly add the 'debug' tag \
         to native targets (both compilation and linking)."
@@ -159,15 +156,13 @@ rule "ocaml: mlpack & cmi & cmx* & o* -> cmx & o"
   ~deps:["%.mlpack"; "%.cmi"]
   ~doc:"If foo.mlpack contains a list of capitalized module names, \
   the target foo.cmx will produce a packed module containing \
-  those modules as submodules.
-
-\
+  those modules as submodules.\n\
+  \n\
   Warning: The .cmx files that will be included must be manually tagged \
   with the tag \"for-pack(foo)\". This means that you cannot include \
   the same bar.cmx in several .mlpack files, and that you should not \
-  use an included .cmx as a separate module on its own.
-
-\
+  use an included .cmx as a separate module on its own.\n\
+  \n\
   This requirement comes from a technical limitation of \
   native module packing: ocamlopt needs the -for-pack argument to be passed \
   ahead of time, when compiling each included submodule, \
@@ -197,9 +192,8 @@ rule "ocaml: ml -> cmo & cmi"
   ~prods:["%.cmo"; "%.cmi"]
   ~deps:["%.ml"; "%.ml.depends"]
   ~doc:"This rule allows to produce a .cmi from a .ml file \
-        when the corresponding .mli is missing.
-
-\
+        when the corresponding .mli is missing.\n\
+        \n\
         Note: you are strongly encourage to have a .mli file \
         for each of your .ml module, as it is a good development \
         practice which also simplifies the way build systems work, \
@@ -408,9 +402,8 @@ rule "ocaml dependencies ml"
   ~dep:"%.ml"
   ~doc:"OCamlbuild will use ocamldep to approximate dependencies \
         of a source file. The ocamldep tool being purely syntactic, \
-        it only computes an over-approximation of the dependencies.
-
-\
+        it only computes an over-approximation of the dependencies.\n\
+        \n\
         If you manipulate a module Foo that is in fact a submodule Bar.Foo \
         (after 'open Bar'), ocamldep may believe that your module depends \
         on foo.ml -- when such a file also exists in your project. This can \
@@ -590,9 +583,8 @@ rule "preprocess: ml -> pp.ml"
   ~prod:"%.pp.ml"
   ~doc:"The target foo.pp.ml should generate a source file equivalent \
         to foo.ml after syntactic preprocessors (camlp4, etc.) have been \
-        applied.
-
-\
+        applied.\n\
+        \n\
         Warning: This option is currently known to malfunction \
         when used together with -use-ocamlfind (for syntax extensions \
         coming from ocamlfind packages). Direct compilation of the \

--- a/src/options.ml
+++ b/src/options.ml
@@ -171,7 +171,7 @@ let set_cmd rcmd = String (fun s -> rcmd := Sh s)
 let set_build_dir s =
   make_links := false;
   if Filename.is_relative s then
-    build_dir := Filename.concat (Sys.getcwd ()) s
+    build_dir := filename_concat (Sys.getcwd ()) s
   else
     build_dir := s
 let spec = ref (

--- a/src/rule.ml
+++ b/src/rule.ml
@@ -269,7 +269,7 @@ let rule name ?tags ?(prods=[]) ?(deps=[]) ?prod ?dep ?stamp ?(insert = `bottom)
     match tags with
       | None -> ()
       | Some _ ->
-        Log.eprintf "Warning: your ocamlbuild rule %S uses the ~tags parameter,
+        Log.eprintf "Warning: your ocamlbuild rule %S uses the ~tags parameter, \
                      which is deprecated and ignored."
           name
   in

--- a/testsuite/external.ml
+++ b/testsuite/external.ml
@@ -6,7 +6,12 @@ let () = test "SyntaxFlag"
   ~options:[`use_ocamlfind; `package "camlp4.macro"; `syntax "camlp4o"]
   ~requirements:(package_exists "camlp4.macro")
   ~description:"-syntax for ocamlbuild"
-  ~tree:[T.f "dummy.ml" ~content:"IFDEF TEST THEN\nprint_endline \"Hello\";;\nENDIF;;"]
+  ~tree:[T.f "dummy.ml" ~content:
+{|
+IFDEF TEST THEN
+print_endline "Hello";;
+ENDIF;;
+|}]
   ~matching:[M.f "dummy.native"]
   ~targets:("dummy.native",[]) ();;
 
@@ -34,11 +39,13 @@ let () = test "SubtoolOptions"
   ~requirements:(req_and (package_exists "menhirLib") (package_exists "camlp4"))
   ~options:[`use_ocamlfind; `use_menhir; `tags ["package(camlp4.fulllib)"]]
   ~tree:[T.f "parser.mly"
-            ~content:"%{ %}\n\
-                      %token DUMMY\n\
-                      %start<Camlp4.PreCast.Syntax.Ast.expr option> test\n\
-                      %%\n\
-                      test: DUMMY {None}"]
+            ~content:{|
+%{ %}
+%token DUMMY
+%start<Camlp4.PreCast.Syntax.Ast.expr option> test
+%%
+test: DUMMY {None}
+|}]
   ~matching:[M.f "parser.native"; M.f "parser.byte"]
   ~targets:("parser.native",["parser.byte"])
   ();;

--- a/testsuite/external.ml
+++ b/testsuite/external.ml
@@ -34,10 +34,10 @@ let () = test "SubtoolOptions"
   ~requirements:(req_and (package_exists "menhirLib") (package_exists "camlp4"))
   ~options:[`use_ocamlfind; `use_menhir; `tags ["package(camlp4.fulllib)"]]
   ~tree:[T.f "parser.mly"
-            ~content:"%{ %}
-                      %token DUMMY
-                      %start<Camlp4.PreCast.Syntax.Ast.expr option> test
-                      %%
+            ~content:"%{ %}\n\
+                      %token DUMMY\n\
+                      %start<Camlp4.PreCast.Syntax.Ast.expr option> test\n\
+                      %%\n\
                       test: DUMMY {None}"]
   ~matching:[M.f "parser.native"; M.f "parser.byte"]
   ~targets:("parser.native",["parser.byte"])

--- a/testsuite/internal.ml
+++ b/testsuite/internal.ml
@@ -184,21 +184,21 @@ let () = test "CmxsStubLink"
            T.f "libfoo_stubs.clib" ~content:"foo_stubs.o";
            T.f "foo.ml" ~content:"";
          ];
-         T.f "_tags" ~content:"
-<src/foo.{cma,cmxa}> : record_foo_stubs
-<src/foo.cmxs> : link_foo_stubs";
-         T.f "myocamlbuild.ml" ~content:"
-open Ocamlbuild_plugin
-let () =
-  dispatch begin function
-  | After_rules ->
-      dep [\"record_foo_stubs\"] [\"src/libfoo_stubs.a\"];
-      flag_and_dep
-        [\"link\"; \"ocaml\"; \"link_foo_stubs\"] (P \"src/libfoo_stubs.a\");
-      flag [\"library\"; \"ocaml\";           \"record_foo_stubs\"]
-        (S ([A \"-cclib\"; A \"-lfoo_stubs\"]));
-  | _ -> ()
-  end
+         T.f "_tags" ~content:"\
+<src/foo.{cma,cmxa}> : record_foo_stubs\n\
+<src/foo.cmxs> : link_foo_stubs\n";
+         T.f "myocamlbuild.ml" ~content:"\
+open Ocamlbuild_plugin\n\
+let () =\n\
+  dispatch begin function\n\
+  | After_rules ->\n\
+      dep [\"record_foo_stubs\"] [\"src/libfoo_stubs.a\"];\n\
+      flag_and_dep\n\
+        [\"link\"; \"ocaml\"; \"link_foo_stubs\"] (P \"src/libfoo_stubs.a\");\n\
+      flag [\"library\"; \"ocaml\";           \"record_foo_stubs\"]\n\
+        (S ([A \"-cclib\"; A \"-lfoo_stubs\"]));\n\
+  | _ -> ()\n\
+  end\n\
 "]
   ~targets:("src/foo.cmxs",[]) ();;
 
@@ -208,31 +208,31 @@ let () = test "StrictSequenceFlag"
   ~tree:[T.f "hello.ml" ~content:"let () = 1; ()";
          T.f "_tags" ~content:"true: strict_sequence\n"]
   ~failing_msg:(if Sys.ocaml_version < "4.07.0" then
-"File \"hello.ml\", line 1, characters 9-10:
-Error: This expression has type int but an expression was expected of type
-         unit
+"File \"hello.ml\", line 1, characters 9-10:\n\
+Error: This expression has type int but an expression was expected of type\n\
+\         unit\n\
 Command exited with code 2."
 else if Sys.ocaml_version < "4.08.0" then
-"File \"hello.ml\", line 1, characters 9-10:
-Error: This expression has type int but an expression was expected of type
-         unit
-       because it is in the left-hand side of a sequence
+"File \"hello.ml\", line 1, characters 9-10:\n\
+Error: This expression has type int but an expression was expected of type\n\
+\         unit\n\
+\       because it is in the left-hand side of a sequence\n\
 Command exited with code 2."
 else if Sys.ocaml_version < "5.2.0" then
-"File \"hello.ml\", line 1, characters 9-10:
-1 | let () = 1; ()
-             ^
-Error: This expression has type int but an expression was expected of type
-         unit
-       because it is in the left-hand side of a sequence
+"File \"hello.ml\", line 1, characters 9-10:\n\
+1 | let () = 1; ()\n\
+\             ^\n\
+Error: This expression has type int but an expression was expected of type\n\
+\         unit\n\
+\       because it is in the left-hand side of a sequence\n\
 Command exited with code 2."
 else
-"File \"hello.ml\", line 1, characters 9-10:
-1 | let () = 1; ()
-             ^
-Error: This expression has type \"int\" but an expression was expected of type
-         \"unit\"
-       because it is in the left-hand side of a sequence
+"File \"hello.ml\", line 1, characters 9-10:\n\
+1 | let () = 1; ()\n\
+\             ^\n\
+Error: This expression has type \"int\" but an expression was expected of type\n\
+\         \"unit\"\n\
+\       because it is in the left-hand side of a sequence\n\
 Command exited with code 2."
 )
   ~targets:("hello.byte",[]) ();;
@@ -243,16 +243,16 @@ let () = test "StrictFormatsFlag"
   ~tree:[T.f "hello.ml" ~content:"let _ = Printf.printf \"%.10s\"";
          T.f "_tags" ~content:"true: strict_formats\n"]
   ~failing_msg:(if Sys.ocaml_version < "4.08.0" then
-"File \"hello.ml\", line 1, characters 22-29:
+"File \"hello.ml\", line 1, characters 22-29:\n\
 Error: invalid format \"%.10s\": at character number 0, \
-`precision' is incompatible with 's' in sub-format \"%.10s\"
+`precision' is incompatible with 's' in sub-format \"%.10s\"\n\
 Command exited with code 2."
 else
-"File \"hello.ml\", line 1, characters 22-29:
-1 | let _ = Printf.printf \"%.10s\"
-                          ^^^^^^^
+"File \"hello.ml\", line 1, characters 22-29:\n\
+1 | let _ = Printf.printf \"%.10s\"\n\
+\                          ^^^^^^^\n\
 Error: invalid format \"%.10s\": at character number 0, \
-`precision' is incompatible with 's' in sub-format \"%.10s\"
+`precision' is incompatible with 's' in sub-format \"%.10s\"\n\
 Command exited with code 2."
 )
   ~targets:("hello.byte",[]) ();;
@@ -261,21 +261,21 @@ let () = test "PrincipalFlag"
   ~options:[`no_ocamlfind; `quiet]
   ~description:"-principal tag"
   ~tree:[T.f "hello.ml"
-            ~content:"type s={foo:int;bar:unit} type t={foo:int}
+            ~content:"type s={foo:int;bar:unit} type t={foo:int}\n\
                       let f x = (x.bar; x.foo)";
          T.f "_tags" ~content:"true: principal\n"]
   ~failing_msg:(if Sys.ocaml_version < "4.08.0" then
-"File \"hello.ml\", line 2, characters 42-45:
+"File \"hello.ml\", line 2, characters 20-23:\n\
 Warning 18: this type-based field disambiguation is not principal."
 else if Sys.ocaml_version < "4.12.0" then
-"File \"hello.ml\", line 2, characters 42-45:
-2 |                       let f x = (x.bar; x.foo)
-                                              ^^^
+"File \"hello.ml\", line 2, characters 20-23:\n\
+2 | let f x = (x.bar; x.foo)\n\
+\                        ^^^\n\
 Warning 18: this type-based field disambiguation is not principal."
 else
-"File \"hello.ml\", line 2, characters 42-45:
-2 |                       let f x = (x.bar; x.foo)
-                                              ^^^
+"File \"hello.ml\", line 2, characters 20-23:\n\
+2 | let f x = (x.bar; x.foo)\n\
+\                        ^^^\n\
 Warning 18 [not-principal]: this type-based field disambiguation is not principal."
 )
   ~targets:("hello.byte",[]) ();;
@@ -296,7 +296,7 @@ let () = test "ModularPlugin2"
   ~tree:[T.f "main.ml" ~content:"let x = 1";
          T.f "_tags" ~content:"<main.*>: toto(-g)";
          T.f "myocamlbuild.ml"
-           ~content:"open Ocamlbuild_plugin;;
+           ~content:"open Ocamlbuild_plugin;;\n\
                      pflag [\"link\"] \"toto\" (fun arg -> A arg);;"]
   ~failing_msg:""
   ~matching:[M.f "main.byte"]
@@ -309,7 +309,7 @@ let () = test "ModularPlugin3"
   ~options:[`no_ocamlfind; `quiet; `plugin_tag "toto(-g)"]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
          T.f "myocamlbuild.ml"
-           ~content:"open Ocamlbuild_plugin;;
+           ~content:"open Ocamlbuild_plugin;;\n\
                      pflag [\"link\"] \"toto\" (fun arg -> A arg);;"]
   ~failing_msg:"Warning: tag \"toto\" does not expect a parameter, \
                 but is used with parameter \"-g\""
@@ -375,12 +375,12 @@ let () = test "TagsNewlines"
   ~tree:[
     T.f "main.ml" ~content:"";
     T.f "_tags" ~content:
-"<foo>: debug,\\
-rectypes
-<bar>: \\
-debug, rectypes
-<baz>\\
-: debug, rectypes
+"<foo>: debug,\\\n\
+rectypes\n\
+<bar>: \\\n\
+debug, rectypes\n\
+<baz>\\\n\
+: debug, rectypes\n\
 ";
   ]
   ~matching:[M.f "main.byte"]
@@ -487,7 +487,7 @@ let () = test "CmxsFromMllib1"
   ~targets:("foo.cmxs", []) ();;
 
 let () = test "CmxsFromMllib2"
-  ~description:"Check that a .cmxs file can be built from a .mllib file,
+  ~description:"Check that a .cmxs file can be built from a .mllib file, \
                 even when one of the module has the same name as the library"
   ~requirements:ocamlopt_available
   ~options:[`no_ocamlfind; `no_plugin]

--- a/testsuite/internal.ml
+++ b/testsuite/internal.ml
@@ -165,13 +165,15 @@ let () = test "OutputObj"
   ~tree:[T.f "hello.ml" ~content:"print_endline \"Hello, World!\""]
   ~targets:("hello.byte.o",["hello.byte.c";"hello.native.o"]) ();;
 
+let so = if Sys.win32 then "dll" else "so"
+
 let () = test "OutputShared"
   ~options:[`no_ocamlfind]
   ~description:"output_shared targets for native and bytecode (PR #6733)"
   ~requirements:ocamlopt_available
   ~tree:[T.f "hello.ml" ~content:"print_endline \"Hello, World!\"";
          T.f "_tags" ~content:"<*.so>: runtime_variant(_pic)"]
-  ~targets:("hello.byte.so",["hello.native.so"]) ();;
+  ~targets:("hello.byte."^so,["hello.native."^so]) ();;
 
 let () = test "CmxsStubLink"
   ~options:[`no_ocamlfind]
@@ -314,12 +316,14 @@ let () = test "ModularPlugin3"
   ~matching:[M.f "main.byte"]
   ~targets:("main.byte",[]) ();;
 
+let exe_suf = if Sys.win32 then ".exe" else ""
+
 let () = test "PluginCompilation1"
   ~description:"check that the plugin is not compiled when -no-plugin is passed"
   ~options:[`no_ocamlfind; `no_plugin]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
          T.f "myocamlbuild.ml" ~content:"prerr_endline \"foo\";;"]
-  ~matching:[_build [M.Not (M.f "myocamlbuild")]]
+  ~matching:[_build [M.Not (M.f ("myocamlbuild" ^ exe_suf))]]
   ~targets:("main.byte",[]) ();;
 
 let () = test "PluginCompilation2"
@@ -327,7 +331,7 @@ let () = test "PluginCompilation2"
   ~options:[`no_ocamlfind; `just_plugin]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
          T.f "myocamlbuild.ml" ~content:"print_endline \"foo\";;"]
-  ~matching:[_build [M.f "myocamlbuild"]]
+  ~matching:[_build [M.f ("myocamlbuild" ^ exe_suf)]]
   ~targets:("", []) ();;
 
 let () = test "PluginCompilation3"

--- a/testsuite/internal.ml
+++ b/testsuite/internal.ml
@@ -53,14 +53,17 @@ let () = test "BuildDir"
   ~targets:("dummy.byte",[]) ();;
 
 let tag_pat_msgs =
-  ["*:a", "File \"_tags\", line 1, characters 0-2:\n\
-           Lexing error: Invalid globbing pattern \"*\".";
+  ["*:a",
+{|File "_tags", line 1, characters 0-2:
+Lexing error: Invalid globbing pattern "*".|};
 
-   "\n<*{>:a", "File \"_tags\", line 2, characters 0-5:\n\
-                Lexing error: Invalid globbing pattern \"<*{>\".";
+   "\n<*{>:a",
+{|File "_tags", line 2, characters 0-5:
+Lexing error: Invalid globbing pattern "<*{>".|};
 
-   "<*>: ~@a,# ~a", "File \"_tags\", line 1, characters 10-11:\n\
-                     Lexing error: Only ',' separated tags are allowed."];;
+   "<*>: ~@a,# ~a",
+{|File "_tags", line 1, characters 10-11:
+Lexing error: Only ',' separated tags are allowed.|}];;
 
 List.iteri (fun i (content,failing_msg) ->
   let () = test (Printf.sprintf "TagsErrorMessage_%d" (i+1))
@@ -74,7 +77,12 @@ List.iteri (fun i (content,failing_msg) ->
 let () = test "Itarget"
   ~options:[`no_ocamlfind]
   ~description:".itarget building with dependencies between the modules (PR#5686)"
-  ~tree:[T.f "foo.itarget" ~content:"a.cma\nb.byte\n"; T.f "a.ml"; T.f "b.ml" ~content:"open A\n"]
+  ~tree:[T.f "foo.itarget" ~content:{|
+a.cma
+b.byte
+|};
+         T.f "a.ml";
+         T.f "b.ml" ~content:"open A"]
   ~matching:[M.f "a.cma"; M.f "b.byte"]
   ~targets:("foo.otarget",[]) ();;
 
@@ -82,9 +90,12 @@ let () = test "PackAcross"
   ~options:[`no_ocamlfind]
   ~description:"Pack using a module from the other tree (PR#4592)"
   ~requirements:ocamlopt_available
-  ~tree:[T.f "main.ml" ~content:"let _ = Pack.Packed.g ()\n";
+  ~tree:[T.f "main.ml" ~content:"let _ = Pack.Packed.g ()";
          T.f "Pack.mlpack" ~content:"pack/Packed";
-         T.f "_tags" ~content:"<lib>: include\n<pack/*.cmx>: for-pack(Pack)\n";
+         T.f "_tags" ~content:{|
+<lib>: include
+<pack/*.cmx>: for-pack(Pack)
+|};
          T.d "lib" [T.f "Lib.ml" ~content:"let f()=()";
                     T.f "Lib.mli" ~content:"val f : unit -> unit"];
          T.d "pack" [T.f "Packed.ml" ~content:"let g() = Lib.f ()"]]
@@ -111,7 +122,7 @@ let () = test "PackAcross3"
          T.f "foo.mlpack" ~content:"foo/Bar";
          T.f "main.ml" ~content:"prerr_endline Foo.Bar.baz";
          T.f "myocamlbuild.ml";
-         T.f "quux.ml" ~content:"let xyzzy = \"xyzzy\"";
+         T.f "quux.ml" ~content:{|let xyzzy = "xyzzy"|};
          T.f "quux.mli" ~content:"val xyzzy : string"]
   ~matching:[M.f "main.byte"]
   ~targets:("main.byte",[]) ();;
@@ -131,7 +142,7 @@ let () = test "NoIncludeNoHygiene1"
   ~description:"check that hygiene checks are only done in traversed directories\
                 (PR#4502)"
   ~tree:[T.d "must_ignore" [ T.f "dirty.mli" ~content:"val bug : int"];
-         T.f "hello.ml" ~content:"print_endline \"Hello, World!\"";
+         T.f "hello.ml" ~content:{|print_endline "Hello, World!"|};
          T.f "_tags" ~content:"<must_ignore>: -traverse"]
   ~pre_cmd:"ocamlc -c must_ignore/dirty.mli"
             (* will make hygiene fail if must_ignore/ is checked *)
@@ -142,7 +153,7 @@ let () = test "NoIncludeNoHygiene2"
   ~description:"check that hygiene checks are not done on the -build-dir \
                 (PR#4502)"
   ~tree:[T.d "must_ignore" [ T.f "dirty.mli" ~content:"val bug : int"];
-         T.f "hello.ml" ~content:"print_endline \"Hello, World!\"";
+         T.f "hello.ml" ~content:{|print_endline "Hello, World!"|};
          T.f "_tags" ~content:""]
   ~pre_cmd:"ocamlc -c must_ignore/dirty.mli"
             (* will make hygiene fail if must_ignore/ is checked *)
@@ -152,7 +163,7 @@ let () = test "NoIncludeNoHygiene3"
   ~options:[`no_ocamlfind; `X "must_ignore"]
   ~description:"check that hygiene checks are not done on excluded dirs (PR#4502)"
   ~tree:[T.d "must_ignore" [ T.f "dirty.mli" ~content:"val bug : int"];
-         T.f "hello.ml" ~content:"print_endline \"Hello, World!\"";
+         T.f "hello.ml" ~content:{|print_endline "Hello, World!"|};
          T.f "_tags" ~content:""]
   ~pre_cmd:"ocamlc -c must_ignore/dirty.mli"
             (* will make hygiene fail if must_ignore/ is checked *)
@@ -162,7 +173,7 @@ let () = test "OutputObj"
   ~options:[`no_ocamlfind]
   ~description:"output_obj targets for native and bytecode (PR #6049)"
   ~requirements:ocamlopt_available
-  ~tree:[T.f "hello.ml" ~content:"print_endline \"Hello, World!\""]
+  ~tree:[T.f "hello.ml" ~content:{|print_endline "Hello, World!"|}]
   ~targets:("hello.byte.o",["hello.byte.c";"hello.native.o"]) ();;
 
 let so = if Sys.win32 then "dll" else "so"
@@ -171,7 +182,7 @@ let () = test "OutputShared"
   ~options:[`no_ocamlfind]
   ~description:"output_shared targets for native and bytecode (PR #6733)"
   ~requirements:ocamlopt_available
-  ~tree:[T.f "hello.ml" ~content:"print_endline \"Hello, World!\"";
+  ~tree:[T.f "hello.ml" ~content:{|print_endline "Hello, World!"|};
          T.f "_tags" ~content:"<*.so>: runtime_variant(_pic)"]
   ~targets:("hello.byte."^so,["hello.native."^so]) ();;
 
@@ -184,76 +195,79 @@ let () = test "CmxsStubLink"
            T.f "libfoo_stubs.clib" ~content:"foo_stubs.o";
            T.f "foo.ml" ~content:"";
          ];
-         T.f "_tags" ~content:"\
-<src/foo.{cma,cmxa}> : record_foo_stubs\n\
-<src/foo.cmxs> : link_foo_stubs\n";
-         T.f "myocamlbuild.ml" ~content:"\
-open Ocamlbuild_plugin\n\
-let () =\n\
-  dispatch begin function\n\
-  | After_rules ->\n\
-      dep [\"record_foo_stubs\"] [\"src/libfoo_stubs.a\"];\n\
-      flag_and_dep\n\
-        [\"link\"; \"ocaml\"; \"link_foo_stubs\"] (P \"src/libfoo_stubs.a\");\n\
-      flag [\"library\"; \"ocaml\";           \"record_foo_stubs\"]\n\
-        (S ([A \"-cclib\"; A \"-lfoo_stubs\"]));\n\
-  | _ -> ()\n\
-  end\n\
-"]
+         T.f "_tags" ~content:{|
+<src/foo.{cma,cmxa}> : record_foo_stubs
+<src/foo.cmxs> : link_foo_stubs
+|};
+         T.f "myocamlbuild.ml" ~content:{|
+open Ocamlbuild_plugin
+let () =
+  dispatch begin function
+  | After_rules ->
+      dep ["record_foo_stubs"] ["src/libfoo_stubs.a"];
+      flag_and_dep
+        ["link"; "ocaml"; "link_foo_stubs"] (P "src/libfoo_stubs.a");
+      flag ["library"; "ocaml"; "record_foo_stubs"]
+        (S ([A "-cclib"; A "-lfoo_stubs"]));
+  | _ -> ()
+  end
+|}]
   ~targets:("src/foo.cmxs",[]) ();;
 
 let () = test "StrictSequenceFlag"
   ~options:[`no_ocamlfind; `quiet]
   ~description:"strict_sequence tag"
   ~tree:[T.f "hello.ml" ~content:"let () = 1; ()";
-         T.f "_tags" ~content:"true: strict_sequence\n"]
+         T.f "_tags" ~content:{|
+true: strict_sequence
+|}]
   ~failing_msg:(if Sys.ocaml_version < "4.07.0" then
-"File \"hello.ml\", line 1, characters 9-10:\n\
-Error: This expression has type int but an expression was expected of type\n\
-\         unit\n\
-Command exited with code 2."
+{|File "hello.ml", line 1, characters 9-10:
+Error: This expression has type int but an expression was expected of type
+         unit
+Command exited with code 2.|}
 else if Sys.ocaml_version < "4.08.0" then
-"File \"hello.ml\", line 1, characters 9-10:\n\
-Error: This expression has type int but an expression was expected of type\n\
-\         unit\n\
-\       because it is in the left-hand side of a sequence\n\
-Command exited with code 2."
+{|File "hello.ml", line 1, characters 9-10:
+Error: This expression has type int but an expression was expected of type
+         unit
+       because it is in the left-hand side of a sequence
+Command exited with code 2.|}
 else if Sys.ocaml_version < "5.2.0" then
-"File \"hello.ml\", line 1, characters 9-10:\n\
-1 | let () = 1; ()\n\
-\             ^\n\
-Error: This expression has type int but an expression was expected of type\n\
-\         unit\n\
-\       because it is in the left-hand side of a sequence\n\
-Command exited with code 2."
+{|File "hello.ml", line 1, characters 9-10:
+1 | let () = 1; ()
+             ^
+Error: This expression has type int but an expression was expected of type
+         unit
+       because it is in the left-hand side of a sequence
+Command exited with code 2.|}
 else
-"File \"hello.ml\", line 1, characters 9-10:\n\
-1 | let () = 1; ()\n\
-\             ^\n\
-Error: This expression has type \"int\" but an expression was expected of type\n\
-\         \"unit\"\n\
-\       because it is in the left-hand side of a sequence\n\
-Command exited with code 2."
+{|File "hello.ml", line 1, characters 9-10:
+1 | let () = 1; ()
+             ^
+Error: This expression has type "int" but an expression was expected of type
+         "unit"
+       because it is in the left-hand side of a sequence
+Command exited with code 2.|}
 )
   ~targets:("hello.byte",[]) ();;
 
 let () = test "StrictFormatsFlag"
   ~options:[`no_ocamlfind; `quiet]
   ~description:"strict_format tag"
-  ~tree:[T.f "hello.ml" ~content:"let _ = Printf.printf \"%.10s\"";
-         T.f "_tags" ~content:"true: strict_formats\n"]
+  ~tree:[T.f "hello.ml" ~content:{|let _ = Printf.printf "%.10s"|};
+         T.f "_tags" ~content:{|
+true: strict_formats
+|}]
   ~failing_msg:(if Sys.ocaml_version < "4.08.0" then
-"File \"hello.ml\", line 1, characters 22-29:\n\
-Error: invalid format \"%.10s\": at character number 0, \
-`precision' is incompatible with 's' in sub-format \"%.10s\"\n\
-Command exited with code 2."
+{|File "hello.ml", line 1, characters 22-29:
+Error: invalid format "%.10s": at character number 0, `precision' is incompatible with 's' in sub-format "%.10s"
+Command exited with code 2.|}
 else
-"File \"hello.ml\", line 1, characters 22-29:\n\
-1 | let _ = Printf.printf \"%.10s\"\n\
-\                          ^^^^^^^\n\
-Error: invalid format \"%.10s\": at character number 0, \
-`precision' is incompatible with 's' in sub-format \"%.10s\"\n\
-Command exited with code 2."
+{|File "hello.ml", line 1, characters 22-29:
+1 | let _ = Printf.printf "%.10s"
+                          ^^^^^^^
+Error: invalid format "%.10s": at character number 0, `precision' is incompatible with 's' in sub-format "%.10s"
+Command exited with code 2.|}
 )
   ~targets:("hello.byte",[]) ();;
 
@@ -261,22 +275,26 @@ let () = test "PrincipalFlag"
   ~options:[`no_ocamlfind; `quiet]
   ~description:"-principal tag"
   ~tree:[T.f "hello.ml"
-            ~content:"type s={foo:int;bar:unit} type t={foo:int}\n\
-                      let f x = (x.bar; x.foo)";
-         T.f "_tags" ~content:"true: principal\n"]
+           ~content:
+{|type s={foo:int;bar:unit} type t={foo:int}
+let f x = (x.bar; x.foo)
+|};
+         T.f "_tags" ~content:{|
+true: principal
+|}]
   ~failing_msg:(if Sys.ocaml_version < "4.08.0" then
-"File \"hello.ml\", line 2, characters 20-23:\n\
-Warning 18: this type-based field disambiguation is not principal."
+{|File "hello.ml", line 2, characters 20-23:
+Warning 18: this type-based field disambiguation is not principal.|}
 else if Sys.ocaml_version < "4.12.0" then
-"File \"hello.ml\", line 2, characters 20-23:\n\
-2 | let f x = (x.bar; x.foo)\n\
-\                        ^^^\n\
-Warning 18: this type-based field disambiguation is not principal."
+{|File "hello.ml", line 2, characters 20-23:
+2 | let f x = (x.bar; x.foo)
+                        ^^^
+Warning 18: this type-based field disambiguation is not principal.|}
 else
-"File \"hello.ml\", line 2, characters 20-23:\n\
-2 | let f x = (x.bar; x.foo)\n\
-\                        ^^^\n\
-Warning 18 [not-principal]: this type-based field disambiguation is not principal."
+{|File "hello.ml", line 2, characters 20-23:
+2 | let f x = (x.bar; x.foo)
+                        ^^^
+Warning 18 [not-principal]: this type-based field disambiguation is not principal.|}
 )
   ~targets:("hello.byte",[]) ();;
 
@@ -284,7 +302,7 @@ let () = test "ModularPlugin1"
   ~description:"test a plugin with dependency on external libraries"
   ~options:[`no_ocamlfind; `quiet; `plugin_tag "use_str"]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
-         T.f "myocamlbuild.ml" ~content:"ignore (Str.quote \"\");;"]
+         T.f "myocamlbuild.ml" ~content:{|ignore (Str.quote "");;|}]
   ~matching:[M.f "main.byte"]
   ~targets:("main.byte",[]) ();;
 
@@ -296,8 +314,10 @@ let () = test "ModularPlugin2"
   ~tree:[T.f "main.ml" ~content:"let x = 1";
          T.f "_tags" ~content:"<main.*>: toto(-g)";
          T.f "myocamlbuild.ml"
-           ~content:"open Ocamlbuild_plugin;;\n\
-                     pflag [\"link\"] \"toto\" (fun arg -> A arg);;"]
+           ~content:{|
+open Ocamlbuild_plugin;;
+pflag ["link"] "toto" (fun arg -> A arg);;
+|}]
   ~failing_msg:""
   ~matching:[M.f "main.byte"]
   ~targets:("main.byte",[]) ();;
@@ -309,10 +329,12 @@ let () = test "ModularPlugin3"
   ~options:[`no_ocamlfind; `quiet; `plugin_tag "toto(-g)"]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
          T.f "myocamlbuild.ml"
-           ~content:"open Ocamlbuild_plugin;;\n\
-                     pflag [\"link\"] \"toto\" (fun arg -> A arg);;"]
-  ~failing_msg:"Warning: tag \"toto\" does not expect a parameter, \
-                but is used with parameter \"-g\""
+           ~content:{|
+open Ocamlbuild_plugin;;
+pflag ["link"] "toto" (fun arg -> A arg);;
+|} ]
+  ~failing_msg:
+{|Warning: tag "toto" does not expect a parameter, but is used with parameter "-g"|}
   ~matching:[M.f "main.byte"]
   ~targets:("main.byte",[]) ();;
 
@@ -322,7 +344,7 @@ let () = test "PluginCompilation1"
   ~description:"check that the plugin is not compiled when -no-plugin is passed"
   ~options:[`no_ocamlfind; `no_plugin]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
-         T.f "myocamlbuild.ml" ~content:"prerr_endline \"foo\";;"]
+         T.f "myocamlbuild.ml" ~content:{|prerr_endline "foo";;|}]
   ~matching:[_build [M.Not (M.f ("myocamlbuild" ^ exe_suf))]]
   ~targets:("main.byte",[]) ();;
 
@@ -330,7 +352,7 @@ let () = test "PluginCompilation2"
   ~description:"check that the plugin is compiled when -just-plugin is passed"
   ~options:[`no_ocamlfind; `just_plugin]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
-         T.f "myocamlbuild.ml" ~content:"print_endline \"foo\";;"]
+         T.f "myocamlbuild.ml" ~content:{|print_endline "foo";;|}]
   ~matching:[_build [M.f ("myocamlbuild" ^ exe_suf)]]
   ~targets:("", []) ();;
 
@@ -340,7 +362,7 @@ let () = test "PluginCompilation3"
   ~requirements:ocamlopt_available
   ~options:[`no_ocamlfind; `quiet; `just_plugin]
   ~tree:[T.f "main.ml" ~content:"let x = 1";
-         T.f "myocamlbuild.ml" ~content:"print_endline \"foo\";;"]
+         T.f "myocamlbuild.ml" ~content:{|print_endline "foo";;|}]
   (* if the plugin were executed we'd get "foo" in failing_msg *)
   ~failing_msg:""
   ~targets:("main.byte", []) ();;
@@ -351,8 +373,7 @@ let () = test "PluginTagsWarning"
   ~options:[`no_ocamlfind; `plugin_tag "use_str"]
   ~tree:[T.f "main.ml" ~content:""]
   ~matching:[_build [M.f "main.cmo"]]
-  ~failing_msg:"Warning: option -plugin-tag(s) has no effect \
-                in absence of plugin file \"myocamlbuild.ml\""
+  ~failing_msg:{|Warning: option -plugin-tag(s) has no effect in absence of plugin file "myocamlbuild.ml"|}
   ~targets:("main.ml", []) ();;
 
 let () = test "TagsInNonHygienic"
@@ -363,7 +384,9 @@ let () = test "TagsInNonHygienic"
   ~tree:[
     T.f "main.ml" ~content:"";
     T.d "deps" [T.f "_tags" ~content:""];
-    T.f "_tags" ~content:"<deps>: not_hygienic\n";
+    T.f "_tags" ~content:{|
+<deps>: not_hygienic
+|};
   ]
   ~matching:[M.f "main.byte"]
   ~targets:("main.byte",[]) ();;
@@ -374,14 +397,14 @@ let () = test "TagsNewlines"
   ~options:[`no_ocamlfind]
   ~tree:[
     T.f "main.ml" ~content:"";
-    T.f "_tags" ~content:
-"<foo>: debug,\\\n\
-rectypes\n\
-<bar>: \\\n\
-debug, rectypes\n\
-<baz>\\\n\
-: debug, rectypes\n\
-";
+    T.f "_tags" ~content:{|
+<foo>: debug,\
+rectypes
+<bar>: \
+debug, rectypes
+<baz>\
+: debug, rectypes
+|};
   ]
   ~matching:[M.f "main.byte"]
   ~targets:("main.byte",[]) ();;
@@ -480,9 +503,12 @@ let () = test "CmxsFromMllib1"
   ~requirements:ocamlopt_available
   ~options:[`no_ocamlfind; `no_plugin]
   ~tree:[
-    T.f "a.ml" ~content:"let a = 1\n";
-    T.f "b.ml" ~content:"let b = true\n";
-    T.f "foo.mllib" ~content:"A\nB\n";
+    T.f "a.ml" ~content:"let a = 1";
+    T.f "b.ml" ~content:"let b = true";
+    T.f "foo.mllib" ~content:{|
+A
+B
+|};
   ]
   ~targets:("foo.cmxs", []) ();;
 
@@ -492,10 +518,14 @@ let () = test "CmxsFromMllib2"
   ~requirements:ocamlopt_available
   ~options:[`no_ocamlfind; `no_plugin]
   ~tree:[
-    T.f "a.ml" ~content:"let a = 1\n";
-    T.f "b.ml" ~content:"let b = true\n";
-    T.f "foo.ml" ~content:"let foo = (A.a, B.b)\n";
-    T.f "foo.mllib" ~content:"A\nB\nFoo\n";
+    T.f "a.ml" ~content:"let a = 1";
+    T.f "b.ml" ~content:"let b = true";
+    T.f "foo.ml" ~content:"let foo = (A.a, B.b)";
+    T.f "foo.mllib" ~content:{|
+A
+B
+Foo
+|};
   ]
   ~targets:("foo.cmxs", []) ();;
 
@@ -515,7 +545,9 @@ let () = test "MldylibOverridesMllib"
   ~tree:[
     T.f "foo.ml";
     T.f "bar.ml";
-    T.f "mylib.mllib" ~content:"Foo\nBar";
+    T.f "mylib.mllib" ~content:{|
+FooBar
+|};
     T.f "mylib.mldylib" ~content:"Foo";
   ]
   ~targets:("mylib.cmxs", []) ();;
@@ -529,7 +561,10 @@ let () = test "MldylibOverridesCmx"
   ~tree:[
     T.f "foo.ml";
     T.f "bar.ml";
-    T.f "foo.mldylib" ~content:"Foo\nBar";
+    T.f "foo.mldylib" ~content:{|
+Foo
+Bar
+|};
   ]
   ~targets:("foo.cmx", ["foo.cmxs"]) ();;
 
@@ -542,7 +577,10 @@ let () = test "MllibOverridesCmx"
   ~tree:[
     T.f "foo.ml";
     T.f "bar.ml";
-    T.f "foo.mllib" ~content:"Foo\nBar";
+    T.f "foo.mllib" ~content:{|
+Foo
+Bar
+|};
   ]
   ~targets:("foo.cmx", ["foo.cmxs"]) ();;
 

--- a/testsuite/ocamlbuild_test.ml
+++ b/testsuite/ocamlbuild_test.ml
@@ -435,6 +435,11 @@ let print_colored header_color header name body_color body =
     (color_code header_color) header name
     (color_code body_color) body
 
+let rec mkdir_p dir =
+  if not (Sys.file_exists dir) then begin
+    mkdir_p (Filename.dirname dir);
+    Unix.mkdir dir 0o750;
+  end
 
 let run ~root =
   let dir = Sys.getcwd () in
@@ -449,7 +454,7 @@ let run ~root =
   let build_tree = Filename.dirname test_tree in
 
   let copy l dest =
-    ignore(Sys.command (Printf.sprintf "mkdir -p %s" dest));
+    mkdir_p dest;
     List.iter (fun f -> ignore(Sys.command (Printf.sprintf "cp %s/%s %s" build_tree f dest))) l
   in
   let ocamlbuild = Printf.sprintf "%s/ocamlbuild.byte" install_bin_dir in

--- a/testsuite/ocamlbuild_test.ml
+++ b/testsuite/ocamlbuild_test.ml
@@ -34,6 +34,7 @@ let exists filename = Sys.file_exists filename
 
 let execute cmd =
   let ic = Unix.open_process_in cmd and lst = ref [] in
+  set_binary_mode_in ic false;
   try while true do lst := input_line ic :: !lst done; assert false
   with End_of_file ->
     let ret_code = Unix.close_process_in ic

--- a/testsuite/ocamlbuild_test.ml
+++ b/testsuite/ocamlbuild_test.ml
@@ -30,9 +30,7 @@ let print_string_list = print_list_com pp_print_string
 let print_string_list_com = print_list_com pp_print_string
 let print_string_list_blank = print_list_blank pp_print_string
 
-let exists filename =
-  try ignore(Unix.stat filename); true
-  with Unix.Unix_error ((Unix.ENOENT),_,_) -> false
+let exists filename = Sys.file_exists filename
 
 let execute cmd =
   let ic = Unix.open_process_in cmd and lst = ref [] in

